Revise blueprint (#544)
lightbend/cloudflow: docs/shared-content-source/docs/modules/develop/pages/blueprints.adoc

include::ROOT:partial$include.adoc[]

In Cloudflow, a streaming application is composed of connected `Streamlet`{empty}s.
Each `Streamlet` defines zero or more inlets and outlets.
To create a working application, we must define how the `Streamlet`{empty}s connect together, using _topics_ as communication channels.
 by declaring which outlet from one `Streamlet` should connect with an inlet of another.

A blueprint specifies the streamlets used in an application and how they connect together.
In practical terms, a blueprint is a text file that uses the HOCON format to declare the streamlets in an application, the topics used as communication channels, and the connections between streamlets and topics.

>  the It must be defined in the file `src/main/blueprint/blueprint.conf` within the file structure of the Cloudflow application _sbt project_.

Reviewer: the It?

>  the It must be defined in the file `src/main/blueprint/blueprint.conf` within the file structure of the Cloudflow application _sbt project_.

Reviewer: ```suggestion
 The blueprint must be defined in the file `src/main/blueprint/blueprint.conf` within the file structure of the Cloudflow application _sbt project_.
```

- The short name gives a name to a streamlet instance.
- The implementation of that instance is indicated by the fully qualified name (FQN) of the streamlet implementation class. 
- Each short name must be unique within the application.

The `short-name` becomes part of the application in the following ways:

- We use this name to refer to the streamlets in the _topics_ section.
- It becomes the name of the component in the application runtime (either on a cluster or on the local sandbox)

=== The `streamlets` Section: An Example 

Lets zoom in on the streamlets section in the example below:

[source,hocon]
----
include::{cloudflow-examples-version}@docsnippets:ROOT:example$akkastreams-scala/src/main/blueprint/blueprint.conf[tag=streamlets]
----

The above example shows four streamlet instances defined in the `com.example` package:

- `sensor-data`
- `metrics`
- `validation`
- `reporter`

Streamlet classes are specified by their fully qualified names:

[source, json]
----
sensor-data = com.example.SensorDataIngress
----

It is possible to declare additional instances of any streamlet class. 
This would be a legal declaration:

> [source, json]

Reviewer 2: Why json? The file content is in hocon.

Author: I was trying to see if the rendering (coloring) was better. Changed back.

// Not sure that this belongs here
// The Streamlets can be defined in Java or Scala, as Java or Scala classes with default constructors.
These classes must be available on the classpath, which can be defined directly in the Cloudflow application sbt project or in any dependency that the project has specified, as you would expect of any sbt project.

The streamlet references assigned in the `streamlets` section can be used in the `topics` section to connect streamlets and topics together, as we explore in the next section.

== The `topics` Section

In the `topics` section, we define the connections of an application.
As we explained in xref:app-development-process.adoc[the Cloudflow development process], streamlets have `inlet`{empty}s and `outlet`{empty}s.

=== Consumers and Producers

In the event-based model used by Cloudflow, `inlet`{empty}s are stream consumers and `outlet`{empty}s are stream producers. 

Consumers and producers do not connect point-to-point to each other. 
Instead, we use topics to provide a communication channel between producers and consumers.

> This model allows for an organic growth of an application over time, as newly implemented processes, modelled as streamlets, can be added to an existing application by connecting them to the list of consumers at the appropriate place in the data flow.

Reviewer 2: ```suggestion
This model allows for an organic growth of an application over time, as newly implemented processes, modeled as streamlets, can be added to an existing application by connecting them to the list of consumers at the appropriate place in the data flow.
```

In the following snippet, we declare a three topics that connect to an external managed Kafka instance.
(Note that the credentials are redacted for obvious reasons)

[source, HOCON]
----
cloudflow {
  topics {
    rides = ${shared-config} {      //<1>
      topic.name = "incoming-rides" //<2>
      consumers = [...]
      producers = [...]
    }
    fares = ${shared-config} {      
      topic.name = "incoming-fares"
      consumers = [...]
      producers = [...]
    }
  }
}
shared-config { //<3>                    
  managed = no                       //<4>
  producer-config = ${kafka-connection-config} {
    # more config can be added here from Kafka ProducerConfig (see org.apache.kafka.clients.producer.ProducerConfi)
  }
  consumer-config = ${kafka-connection-config} {
    # more config can be added here from Kafka ConsumerConfig (see org.apache.kafka.clients.consumer.ConsumerConfig)
  }
}
----
<1> Creates a topic definition by extending the `shared-config` with additional configuration.

> <2> Set the topic name explicitly, using the `topic.name` configuration key.

Reviewer 2: ```suggestion
<2> Sets the topic name explicitly, using the `topic.name` configuration key.
```

<3> Adds an arbitrary configuration section to the blueprint definition.
<4> Sets the `managed` topic flag to false.

In the snippet above, we are centralizing the common configuration for all topics in a `shared-config` configuration entry.
We then use the HOCON configuration extension syntax, as in `${shared-config} {...}` to extend the common configuration with the specific values required for each topic definition.

We are also using the explicit `topic.name` property instead of using the default name of the configuration object.

It is important to note the presence of the flag `managed`, which indicated whether Cloudflow should create the topic (`managed=true` (default)) or not (`managed=false`).

> We can setup the Kafka provider configuration in a separate configuration file, like this:

Reviewer 2: ```suggestion
We can set up the Kafka provider configuration in a separate configuration file, like this:
```
